crates/ruff/src/checkers/ast/mod.rs

```diff
@@ -178,26 +188,29 @@ where
 
         // Track whether we've seen docstrings, non-imports, etc.
         match &stmt.node {
-            StmtKind::ImportFrom(ast::StmtImportFrom { module, .. }) => {
+            StmtKind::ImportFrom(ast::StmtImportFrom { module, names, .. }) => {
                 // Allow __future__ imports until we see a non-__future__ import.
-                if self.ctx.futures_allowed {
-                    if let Some(module) = module {
-                        if module != "__future__" {
-                            self.ctx.futures_allowed = false;
-                        }
+                if let Some("__future__") = module.as_deref() {
+                    if names
+                        .iter()
+                        .any(|alias| alias.node.name.as_str() == "annotations")
+                    {
+                        self.ctx.flags |= ContextFlags::FUTURE_ANNOTATIONS;
                     }
+                } else {
+                    self.ctx.flags |= ContextFlags::FUTURES_BOUNDARY;
```

Comment:
`FUTURES_BOUNDARY` seems new. What's the reason for introducing it?

Reply:
It's the same as `futures_allowed`, but renamed for consistency with `IMPORT_BOUNDARY` -- have we seen the future-imports boundary? Or are future imports still ok? (Will add docs for all of these as per above.)